utilities/xmlrpcpp/src/XmlRpcDispatch.cpp

```diff
@@ -81,83 +86,69 @@ XmlRpcDispatch::work(double timeout)
-    }
+    int nEvents = poll(fds, _sources.size(), (timeout_ms < 0) ? -1 : timeout_ms);
 
     if (nEvents < 0)
     {
       if(errno != EINTR)
-        XmlRpcUtil::error("Error in XmlRpcDispatch::work: error in select (%d).", nEvents);
+        XmlRpcUtil::error("Error in XmlRpcDispatch::work: error in poll (%d).", nEvents);
       _inWork = false;
       return;
     }
 
     // Process events
-    for (it=_sources.begin(); it != _sources.end(); )
+    for (i=0, it=_sources.begin(); it != _sources.end(); ++i)
     {
       SourceList::iterator thisIt = it++;
       XmlRpcSource* src = thisIt->getSource();
-      int fd = src->getfd();
       unsigned newMask = (unsigned) -1;
-      if (fd <= maxFd) {
-        // If you select on multiple event types this could be ambiguous
-        if (FD_ISSET(fd, &inFd))
-          newMask &= src->handleEvent(ReadableEvent);
-        if (FD_ISSET(fd, &outFd))
-          newMask &= src->handleEvent(WritableEvent);
-        if (FD_ISSET(fd, &excFd))
-          newMask &= src->handleEvent(Exception);
-
-        // Find the source again.  It may have moved as a result of the way
-        // that sources are removed and added in the call stack starting
-        // from the handleEvent() calls above.
-        for (thisIt=_sources.begin(); thisIt != _sources.end(); thisIt++)
-        {
-          if(thisIt->getSource() == src)
-            break;
-        }
-        if(thisIt == _sources.end())
-        {
-          XmlRpcUtil::error("Error in XmlRpcDispatch::work: couldn't find source iterator");
-          continue;
-        }
-
-        if ( ! newMask) {
-          _sources.erase(thisIt);  // Stop monitoring this one
-          if ( ! src->getKeepOpen())
-            src->close();
-        } else if (newMask != (unsigned) -1) {
-          thisIt->getMask() = newMask;
-        }
+      // If you select on multiple event types this could be ambiguous
+      if (fds[i].revents & POLLIN)
+        newMask &= src->handleEvent(ReadableEvent);
+      if (fds[i].revents & POLLOUT)
+        newMask &= src->handleEvent(WritableEvent);
+      if (fds[i].revents & (POLLERR|POLLHUP|POLLNVAL))
+        newMask &= src->handleEvent(Exception);
+
+      // Find the source again.  It may have moved as a result of the way
+      // that sources are removed and added in the call stack starting
+      // from the handleEvent() calls above.
+      for (thisIt=_sources.begin(); thisIt != _sources.end(); thisIt++)
+      {
+        if(thisIt->getSource() == src)
+          break;
+      }
+      if(thisIt == _sources.end())
+      {
+        XmlRpcUtil::error("Error in XmlRpcDispatch::work: couldn't find source iterator");
+        continue;
+      }
+
+      if ( ! newMask) {
+        _sources.erase(thisIt);  // Stop monitoring this one
+        if ( ! src->getKeepOpen())
+          src->close();
+      } else if (newMask != (unsigned) -1) {
+        thisIt->getMask() = newMask;
```

Comment:
Since we're mutating these lines anyway, perhaps switch to ROS style for the braces? http://wiki.ros.org/CppStyleGuide#Formatting